cli/utils.go

```diff
@@ -29,3 +37,94 @@ func getMapString(m map[string]any, key string) string {
 	}
 	return ""
 }
+
+// replaceInTarGz adds entries to a .tar.gz file by the awkward method of copying
+// its contents to a new file.
+func replaceInTarGz(path string, newEntries map[string][]byte) error {
+	tmpPath := path + ".tmp"
+	// nested function so the Close() calls finish before we rename.
+	err := func() error {
+		// open reader
+		sourceFile, err := os.Open(path) //nolint:gosec
+		if err != nil {
+			return err
+		}
+		defer sourceFile.Close() //nolint:errcheck
+		gzRead, err := gzip.NewReader(sourceFile)
+		if err != nil {
+			return err
+		}
+		defer gzRead.Close() //nolint:errcheck
+		tarRead := tar.NewReader(gzRead)
+
+		// open writer
+		destFile, err := os.Create(tmpPath) //nolint:gosec
+		if err != nil {
+			return err
+		}
+		defer destFile.Close() //nolint:errcheck
+		gzw := gzip.NewWriter(destFile)
+		tarWrite := tar.NewWriter(gzw)
+
+		consumed := make(map[string]bool)
+		// copy existing files, using newEntries where necessary
+		for {
+			head, err := tarRead.Next()
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			if err != nil {
+				return err
+			}
+
+			if entry, ok := newEntries[head.Name]; ok {
+				head.Size = int64(len(entry))
+				if err := tarWrite.WriteHeader(head); err != nil {
+					return err
+				}
+				if _, err := tarWrite.Write(entry); err != nil {
+					return err
+				}
+				consumed[head.Name] = true
+			} else {
+				if err := tarWrite.WriteHeader(head); err != nil {
+					return err
+				}
+				if _, err := io.Copy(tarWrite, tarRead); err != nil { //nolint:gosec
+					return err
+				}
+			}
+		}
+
+		// copy any remaining newEntries
+		for name, entry := range newEntries {
+			if consumed[name] {
+				continue
+			}
+			if err := tarWrite.WriteHeader(&tar.Header{
+				Typeflag: tar.TypeReg,
+				Name:     name,
+				Size:     int64(len(entry)),
+				ModTime:  time.Now(),
+			}); err != nil {
+				return err
+			}
+			if _, err := tarWrite.Write(entry); err != nil {
+				return err
+			}
+		}
+
+		// cleanup
+		if err := tarWrite.Close(); err != nil {
+			return err
+		}
+		if err := gzw.Close(); err != nil {
+			return err
+		}
+		return nil
+	}()
+	if err != nil {
+		return err
+	}
```

Comment:
any reason all of the prior logic is wrapped in a function if we are just going to return the err directly?

I suspect you did this for cleanup reasons but I don't see it.

Reply:
I wanted to make sure `defer sourceFile.Close()` ran before the rename. in practice I'm not sure this matters, but it's the kind of thing that can be platform specific so I don't trust my testing to be general

should I comment / am I overthinking this?

Reply:
Ah -- that makes sense. LGTM.

A comment could be good. But I also doubt anyone will change this.

Reply:
commented in 898070b7d